charms/worker/k8s/src/charm.py

```diff
@@ -1310,14 +1323,124 @@ def _ensure_cert_sans(self):
                 "%s not in cert SANs. Refreshing certs with new SANs: %s", missing_sans, all_sans
             )
             status.add(ops.MaintenanceStatus("Refreshing Certificates"))
-            if self.config.get("bootstrap-certificates") == "self-signed":
+            if self.config.get(CONFIG_BOOTSTRAP_CERTIFICATES) == "self-signed":
                 self.api_manager.refresh_certs(all_sans)
-            elif self.config.get("bootstrap-certificates") == "external":
+            elif self.config.get(CONFIG_BOOTSTRAP_CERTIFICATES) == "external":
                 self.certificate_refresh.emit()
             log.info("Certificates have been refreshed")
 
         log.info("Certificate SANs are up-to-date")
 
+    @on_error(
+        ops.BlockedStatus("Cannot change bootstrap config after bootstrap. Check logs."),
+        BootstrapConfigChangeError,
+    )
+    @on_error(
+        ops.ErrorStatus("Prevent bootstrap config change: Failed to get cluster config"),
+        InvalidResponseError,
+        K8sdConnectionError,
+    )
+    def _prevent_bootstrap_config_change(self):
+        """Prevent bootstrap config changes after bootstrap."""
+        log.info("Preventing bootstrap config changes after bootstrap")
+
+        try:
+            cluster_config = bootstrap_config_options_from_cluster_config(
+                self.api_manager.get_cluster_config().metadata
+            )
+        except InvalidResponseError as e:
+            if e.code == HTTP_503_SERVICE_UNAVAILABLE:
+                log.info("k8sd is not ready, skipping bootstrap config check")
+                return
+            raise
+
+        # NOTE(Hue): The way we're saving certificates has nothing to do with the snap
+        # so we can't rely on the snap to tell us if the certificates have changed.
+        self._stored.set_default(certificates=self.config.get(CONFIG_BOOTSTRAP_CERTIFICATES))
```

Comment:
so we're back to using stored-state?

Reply:
We can infer this from the certificates in the unit:
```python
def check_ca_key():
    return os.path.isFile("/etc/kubernetes/pki/ca.key")
```
This way we can avoid using stored state